llvm/lib/Transforms/IPO/DeadArgumentElimination.cpp

```diff
@@ -1268,5 +1269,70 @@ void DeadArgumentEliminationPass::UpdateNVPTXMetadata(Module &M, Function *F,
       continue;
     // Update the metadata with the new function
     MetadataNode->replaceOperandWith(0, llvm::ConstantAsMetadata::get(NF));
+
+    // Carefully update any and all grid_constant annotations, since those are
+    // denoted parameter indices, which may have changed
+    for (unsigned i = 1; i < MetadataNode->getNumOperands() - 1; i += 2) {
+      if (auto *Type = dyn_cast<MDString>(MetadataNode->getOperand(i));
+          Type && Type->getString() == "grid_constant") {
+        LLVMContext &Ctx = NF->getContext();
+        LLVM_DEBUG(dbgs() << "DeadArgumentEliminationPass - updating nvvm "
+                             "grid_constant annotations for fn: "
+                          << NF->getName() << "\n");
+        // The 'value' operand is a list of integers denoting parameter indices
+        auto *OldGridConstParamIdxs =
+            dyn_cast<MDNode>(MetadataNode->getOperand(i + 1));
+        if (!OldGridConstParamIdxs)
+          continue;
```

Comment:
Do we really expect a case when`!"grid_constant"` is not followed by an `MDNode`? If not, perhaps we could turn this into a regular `cast` to drop runtime check?

Reply:
> I'll change to a regular cast, might be better than silently doing nothing?

Note that it would very likely mean a segfault in release build due to `nullptr` dereference. I'm fine with either approach: for example, you can have an unreachable under `#ifndef NDEBUG` to catch unexpected situations in local development/CI, but still be able to progress further in release builds. But that's once again won't help us to detect an unexpected LLVM IR somewhere in users codebase, because they won't see any diagnostics at all.

Reply:
I've gone with an `assert`, thanks.